runtime/runtime_impl_jetty12/src/main/java/com/google/apphosting/runtime/jetty/http/JettyHttpHandler.java

```diff
@@ -108,8 +110,25 @@ public boolean handle(Request request, Response response, Callback callback) thr
     ApiProxy.Environment currentEnvironment = ApiProxy.getCurrentEnvironment();
     request.setAttribute(AppEngineConstants.ENVIRONMENT_ATTR, currentEnvironment);
 
+    // Only run code to finish request with the RequestManager once the stream is complete.
+    request.addHttpStreamWrapper(
```

Comment:
use addCompletionListener instead, as it has a slightly optimized implementation if multiple completion listeners are registered (which they will be if sessions are used).